Add drag and drop export variables

fkeyzuwu opened the pull request:
Duplicate of https://github.com/godotengine/godot/pull/106186, had some git trouble and couldn't figure how to revert 😅
Closes [#12394](https://github.com/godotengine/godot-proposals/issues/12394)

Think this is pretty much ready for merge on my end. Was reviewed already a bit in the older pr.

@MarianoGnu

fkeyzuwu committed:
Add drag n drop export variables

MarianoGnu commented:
Thx for mention but i dont have rights to approve :)

kitbdev commented on editor/plugins/script_text_editor.cpp:
```diff
@@ -2355,6 +2432,13 @@ void ScriptTextEditor::drop_data_fw(const Point2 &p_point, const Variant &p_data
 
+	// assign dragged exported variables after script chanes
+	if (dragged_export_refs.size() > 0) {
+		connect("edited_script_changed",
```
`edited_script_changed` isn't the right place for this. If you set the Script completion idle parse delay very high (in editor settings) and then save, you can see that the exported variable will appear in the inspector, but it won't be set until the idle parse happens.
I think `_assign_dragged_export_variables` needs to be called from `apply_code` and `_validate_script` instead.

fkeyzuwu replied:
Yeah, tested it and it seems you are right, it works better with any type of configuration I tried tackling it with.

One thing though is I had some bug where the editor crashed after changing the idle parse delay in between the export variables updating (changing from high to low). Couldn't recreate it though.

kitbdev reviewed (commented):
All of the GDScript specific code will need to be abstracted out in the future, but it matches the current behavior so its fine for now.
I don't like how this yet adds another mode where there is no indication that it is possible until after it is dropped. But that existed before, so it can be fixed separately.

fkeyzuwu commented:
> All of the GDScript specific code will need to be abstracted out in the future, but it matches the current behavior so its fine for now. I don't like how this yet adds another mode where there is no indication that it is possible until after it is dropped. But that existed before, so it can be fixed separately.

Yup, that's what I was afraid of, doing too much things that aren't related to the general script text editor, but rather to the GDScript text editor specifically. That's why I attempted to add as little as possible to the actual class structure, like the suggested DraggedExport struct and class level variable. But if you say this is fine, I will change the implementation. Thanks for the review!

fkeyzuwu commented on editor/plugins/script_text_editor.cpp:
```diff
@@ -2233,40 +2260,69 @@ void ScriptTextEditor::drop_data_fw(const Point2 &p_point, const Variant &p_data
+				return;
+			}
+
+			if (!ClassDB::is_parent_class(script->get_instance_base_type(), "Node")) {
+				EditorToaster::get_singleton()->popup_str(vformat(TTR("Can't assign @export variables because script '%s' does not inherit Node."), get_name()), EditorToaster::SEVERITY_WARNING);
```
How would it make sense specifically for resources? Since if you are drag n dropping onto a resource script, we would need an instance of the resource if we want to set it to something. And even then its only one resource instance. Not really sure how much it makes sense. I'm not opposed to still making creating the `@export` line, and just not assigning anything to it afterwards.

MarianoGnu replied:
asigning the value to every single instance existing is a bad idea, is very probablt not what the user wants to do, `EditorInterface.get_inspector().get_edited_object()` gives you the object that is being edited in the inspector, if that object happens to be a Resource, and that Resource happens to have assigned this script, then would make sense to assign the value, otherwise i would not want to assign a value to anything, because will cause in most of the cases the need of the user to go to the resources one by one to clear it

MarianoGnu commented:
> 1. I opted to use mark_scene_as_unsaved instead of using undo-redo, since there can pretty easily be a situation where the undo does not do anything if the @export line was deleted/changed from the script.

- Create an export variable
- Go to the scene and assign it
- Go to the script and delete the line
- Go to scene view (2d or 3d) and undo, output will some something like "Undo: set somevar" this is because the uno simply uses the `set(property, value)` method, and the property could simply be ignored

kitbdev commented on editor/plugins/script_text_editor.cpp:
```diff
@@ -2370,6 +2444,29 @@ void ScriptTextEditor::drop_data_fw(const Point2 &p_point, const Variant &p_data
+
+	ERR_FAIL_COND(sn->get_script_instance() == nullptr);
+
+	for (auto &dragged_export : dragged_exports_dirty) {
```
```suggestion
	for (const DraggedExport &dragged_export : dragged_exports_dirty) {
```
We try to avoid auto.

kitbdev commented on editor/plugins/script_text_editor.h:
```diff
@@ -238,6 +238,16 @@ class ScriptTextEditor : public ScriptEditorBase {
 
+	struct DraggedExport {
+		String variable_name;
+		Variant ref;
```
`ref` kind of implies it is a Ref type, this could be `new_value` or something.

fkeyzuwu replied:
I think just maybe `value` makes more sense, since it isn't its like a value is replaced by an older one

kitbdev commented on editor/plugins/script_text_editor.cpp:
```diff
@@ -2370,6 +2444,29 @@ void ScriptTextEditor::drop_data_fw(const Point2 &p_point, const Variant &p_data
+		sn->get_script_instance()->set(dragged_export.variable_name, dragged_export.ref);
+	}
+
+	EditorInterface::get_singleton()->mark_scene_as_unsaved();
```
We should check the return of `set()` to know if anything was changed, and only set this if there was a change.

fkeyzuwu commented:
> > 1. I opted to use mark_scene_as_unsaved instead of using undo-redo, since there can pretty easily be a situation where the undo does not do anything if the @export line was deleted/changed from the script.
> 
> This can easily happen anyways because the undo history of scripts and scenes are separated
> 
>     * Create an export variable
> 
>     * Go to the scene and assign it
> 
>     * Go to the script and delete the line
> 
>     * Go to scene view (2d or 3d) and undo, output will some something like "Undo: set somevar" this is because the uno simply uses the `set(property, value)` method, and the property could simply be ignored

Yeah, I understand. I just don't see the value in it that much and it also seems a little weird since its only undoing half an action. Not sure about it,

fkeyzuwu commented:
Fixed everything mentioned except undo redo. If it has support from you guys I don't mind implementing it.

Still have a bug because of checking `scene_root->get_script()` and regarding trying to assign variable when having multiple script instances of the same script in one scene. I have some ideas on how to handle it, working on it.

kitbdev commented on editor/plugins/script_text_editor.cpp:
```diff
@@ -2370,6 +2446,72 @@ void ScriptTextEditor::drop_data_fw(const Point2 &p_point, const Variant &p_data
+
+	bool export_variable_set = false;
+	for (const DraggedExport &dragged_export : pending_dragged_exports) {
+		bool success = dragged_export.script_instance->set(dragged_export.variable_name, dragged_export.value);
```
```suggestion
		if (!dragged_export.script_instance) {
			return;
		}
		bool success = dragged_export.script_instance->set(dragged_export.variable_name, dragged_export.value);
```

This needs to make sure the script instance is still valid, since the user can close the scene before it is assigned and cause a crash.
Actually, I'm not sure if just checking that it is nullptr is enough, since it isn't actually ever set to nullptr. You may need to store the ObjectID instead and then get the object from it and get the script instance on that.

fkeyzuwu commented:
Switched to use `ObjectID` to keep track of which `ScriptInstance *`s we need to assign to, instead of just storing raw `ScriptInstance *`. Also fixed everything else, plus added warnings in case the assignment didn't go through in case of scene closed/object deletion/script detachment before script was updated in idle parse.

kitbdev reviewed (approved):
Looks good to me.

Note that this doesn't work on tool scripts, but they do things differently and don't immediately reload in the inspector, so it is fine for this PR.

nubels commented:
I really like this feature. I could see myself using this this over onready vars all the time. However, I've noticed an issue where the assignment of the export variable doesn't work when there are errors in the script. For drag and drop of onready vars it doesn't make a difference if there are errors in the script. I don't think this is that big of a deal though.

I think a good addition to this feature would be an option to create an exported array, say when holding shift + alt. For example, when you have many markers in a scene and drag and drop them into the script using shift + alt, it creates a populated Array[Marker2D]. I don't this this functionality exists for onready vars.

fkeyzuwu commented:
> I really like this feature. I could see myself using this this over onready vars all the time. However, I've noticed an issue where the assignment of the export variable doesn't work when there are errors in the script. For drag and drop of onready vars it doesn't make a difference if there are errors in the script. I don't think this is that big of a deal though.

Yes, that also bothers me a bit. However it goes a bit outside the scope of this PR imo. My idea for this fix is to keep track of export variables in general, in order to handle all kinds of problems with them losing references. So for example the problem you mentioned with drag and dropping while the script has an error, but I also noticed this happens when cutting and pasting the line back in somewhere else in the script, or when renaming the variable. These can also happen one after the other, WHILE the script has errors. So while I can fix that one problem for when the script has errors, I don't think that's the root of the problem. 

> I think a good addition to this feature would be an option to create an exported array, say when holding shift + alt. For example, when you have many markers in a scene and drag and drop them into the script using shift + alt, it creates a populated Array[Marker2D]. I don't this this functionality exists for onready vars.

I also like this, however this also adds functionality for `@onready` vars as well so I would think I would rather add this as a separate PR for both `@export` and `@onready`.

Two really good points though, I've thought about those too.

fkeyzuwu commented:
Just rebased to make sure everything works 🙂

KoBeWi commented on editor/plugins/script_text_editor.cpp:
```diff
@@ -2199,6 +2210,30 @@ static String _get_dropped_resource_line(const Ref<Resource> &p_resource, bool p
+
+	if (resource_script.is_valid()) {
+		StringName global_resource_script_name = resource_script->get_global_name();
+		if (global_resource_script_name != StringName()) {
```
```suggestion
		if (!global_resource_script_name.is_empty()) {
```

fkeyzuwu replied:
Oh, I remember I checked and is_empty() used to check for the length of the string instead of comparing the pointer, which is slower, but it seems that is changed now. Will fix.

KoBeWi reviewed (approved):
Looks fine overall.

Though it's another hidden modifier, it would be nice if it was displayed somewhere. Maybe we could even have a preview of dropped line? idk
(that's something for later)

fkeyzuwu commented:
> Looks fine overall.
> 
> Though it's another hidden modifier, it would be nice if it was displayed somewhere. Maybe we could even have a preview of dropped line? idk (that's something for later)

Yeah, that could be cool 🙂
Also, fixed all `StringName` comparisions in `drop_data_fw()` with `!is_empty()`.

Calinou reviewed (approved):
Tested locally, it works as expected. Code looks good to me.

I agree this needs better discoverability in the future, probably by displaying a hint in the top-left corner of the script editor like we have in the 2D editor already:

<img width="562" height="164" alt="image" src="https://github.com/user-attachments/assets/a3c35cc4-3d54-4d9f-923f-6cf943b6e6a4" />

fkeyzuwu commented:
What do you guys think about updating the selection of the inspector to be the node with the script that the @export variable was assigned to after dropping the line/s? Might be more intuitive to show the user what happened.

MarianoGnu commented:
i dislike the idea of performing actions that wasn't done by the user.
As an user i appreciate software to do what i tell it to do, and not take their own desitions.

Repiteo commented:
Thanks!

SilverWolveGames commented:
Thank you !
